pkg/loop/internal/pb/chain_reader.proto

```diff
@@ -34,3 +77,142 @@ message BoundContract {
   string name = 2;
   bool pending = 3;
 }
+
+enum ComparisonOperator {
+  Eq = 0;
+  Neq = 1;
+  Gt = 2;
+  Lt = 3;
+  Gte = 4;
+  Lte = 5;
+}
+
+message ValueComparator {
+  // TODO pass interface{} here
+  string value = 1;
+  ComparisonOperator operator = 2;
+}
+
+message KeyValuesComparator{
+  string key = 1;
+  repeated ValueComparator value_comparators = 2;
+}
+
+// QueryFilter represents a lightweight orm like DSL defined for filtering over common blockchain primitives.
+message QueryFilter {
+  repeated Expression expression = 1;
+}
+
+message KeyFilter {
+  string key = 1;
+  QueryFilter query_filter = 2;
+}
+
+// Expression encapsulates a single unit of filtering logic, which can be a common blockchain primitive or a composite of boolean expressions. 
+// This allows for both simple and more complex nested expressions.
+message Expression {
+  oneof evaluator {
+    Primitive primitive = 1;
+    BooleanExpression boolean_expression = 2;
+  }
+}
```

Comment:
This is really nice by the way!

I had thought we'd have to resort to just a generic json blob to be able to serialize a DSL tree like this, but this is much better. Didn't realize protobuf was this powerful, fits our needs here perfectly, well done!